File: assets/js/selectors/status.ts
import selectors from 'selectors'
import { RootState } from 'state'
import * as entities from 'state/entities'
import { Focus } from 'state/presence'

export function getById(
  state: RootState,
  id: string
): entities.Status | undefined {
  return state.entities.statuses[id]
}

export function getByUserId(state: RootState, userId: string): entities.Status {
  const status = getById(state, state.presence.userStatuses[userId])
  if (status) {
    return status
  }

  const room = selectors.rooms.getDefaultRoom(state)
  return {
    id: 'default',
    user_id: userId,
    room_id: room?.id || '',
    status: Focus.status,
    camera_on: false,
    mic_on: false,
    speaker_on: false,
    message: '',
    // @ts-ignore
    inserted_at: undefined,
  }
}

export function getLocalStatus(state: RootState): entities.Status {
  const userId = selectors.users.getSelf(state)?.id
  return getByUserId(state, userId || '')
}

File: lib/sway/statuses/status.ex
defmodule Sway.Statuses.Status do
  use Ecto.Schema
  import Ecto.Changeset

  schema "statuses" do
    field :message, :string
    field :status, Ecto.Enum, values: [:online, :focus, :zen]
    field :camera_on, :boolean, default: false
    field :mic_on, :boolean, default: false
    field :speaker_on, :boolean, default: false
    field :timezone, :string, default: nil
    field :emoji, :string, default: nil

    belongs_to :user, Sway.Accounts.User
    belongs_to :room, Sway.Rooms.Room
    belongs_to :workspace, Sway.Workspaces.Workspace

    timestamps()
  end

  @doc false
  def changeset(status, attrs) do
    status
    |> cast(attrs, [:status, :message, :user_id, :room_id, :workspace_id, :camera_on, :mic_on, :speaker_on, :timezone, :emoji])
    |> validate_required([:status, :user_id, :room_id, :workspace_id])
  end
end
